### ISSUE TITLE
EventLoopScheduler Scheduling Bug

### ISSUE THREAD

**Issue by davidwdan** [ReactiveX/RxPHP#36]
Actions that are scheduled from within a scheduled action can be called out of order, which causes unpredictable behavior.

**Comment by asm89** [comment]
Got a code sample that shows the behavior?

**Comment by mbonneau** [comment]
This issue is related to this issue: https://github.com/reactphp/event-loop/issues/14

After taking a closer look, this could be resolved a different way than the existing #37.

Example that reproduces the behavior for me:

``` php
$loop = \React\EventLoop\Factory::create();
$scheduler = new \Rx\Scheduler\EventLoopScheduler($loop);

$o = \Rx\Observable::range(0, 10000)
    ->delay(200)
    ->flatMap(function ($x) {
        return \Rx\Observable::just($x);
    });

$lastItem = -1;
$itemCount = 0;

$o
    ->subscribeCallback(
        function ($x) use (&$lastItem, &$itemCount) {
            echo $x . "\n";
            if ($x !== $lastItem + 1) {
                throw new Exception("Out of order after " . $itemCount . " items");
            }
            $lastItem = $x;
            $itemCount++;
        },
        function ($e) {},
        function () use (&$itemCount) {
            echo "Completed with " . $itemCount . " items.\n";
        },
        $scheduler
    );

$loop->run();
```

The items quickly fall out of order (or possibly never happen because onCompletes are getting called prior to finishing onNexts). There are a few other examples, but they are caused by scheduled events that are not happening in the order expected.

PR #37 fixes the problem by grouping all actions scheduled during a scheduled action together with any other items scheduled for the same time. It stores them in an array and executes them in order at 1 scheduled time.

This can also be resolved by only scheduling one item at a time (using a recursive scheduler if you need to schedule more than one) and always scheduling each item separately (don't call onNext and onCompleted in the same scheduled action). I am not sure if they ran into this issue in RxJS, but it appears that they are generally following these rules. In https://github.com/Reactive-Extensions/RxJS/blob/master/src/core/linq/observable/delay.js they materialize then timestamp recursively scheduling one item at a time using relative times. This option seems like it could be a better solution, but the guidelines would have to be followed when implementing things in the future (or if someone decides to use the scheduler). This only effects operators that schedule actions or create observables. Operators that are just pass-through (like map, etc.) would not need to worry about it.

Most of the project is already compliant with these guidelines (ReturnObservable, delay, and possibly others are not right now).

**Comment by alexsup** [comment]
Here is a quick bit of code that triggers it:

```
$startTime = time();

$loop = \EventLoop\getLoop();

$scheduler = new Rx\Scheduler\EventLoopScheduler($loop);

$observable = new Rx\Observable\ArrayObservable(range(1, 10), $scheduler);

$selectManyObservable = $observable

    ->flatMap(function($value) {
        return  \Rx\React\Http::get("https://publicdata-airports.firebaseio.com/.json");
        
    }, $scheduler)

    ->flatMap(function($value) {
        $nodes = json_decode($value);

        $relatedNodes = [];
        foreach($nodes as $key => $node)
        {
            $relatedNodes[] = $key;
        }

        return new Rx\Observable\ArrayObservable($relatedNodes);

    }, $scheduler)

    ->flatMap(function($value){
        return \Rx\React\Http::get("https://publicdata-airports.firebaseio.com/{$value}.json");
    },$scheduler)


    ->map(function($value) use($startTime) {
        return time() - $startTime;
    },$scheduler);


$disposable = $selectManyObservable->subscribe($stdoutObserver, $scheduler);

$loop->run();

```
